### PR TITLE
Refunding single unit

### PR DESCRIPTION
**Pull request by Zales0123**
zrzut ekranu 2018-07-06 o 11 33 38

### CODE REVIEW

File: src/Provider/RepositoryRefundedUnitTotalProvider.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Provider;
+
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Webmozart\Assert\Assert;
+
+final class RepositoryRefundedUnitTotalProvider implements RefundedUnitTotalProviderInterface
```
Comment (bartoszpietrzak1994) [inline src/Provider/RepositoryRefundedUnitTotalProvider.php]: Honestly, by this class name, I am unable to tell what it exactly does. I guess that the `repository` word is unnecessary.

File: src/Provider/RepositoryOrderRefundedTotalProvider.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Provider;
+
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\RefundPlugin\Entity\RefundInterface;
+
+final class RepositoryOrderRefundedTotalProvider implements OrderRefundedTotalProviderInterface
```
Comment (bartoszpietrzak1994) [inline src/Provider/RepositoryOrderRefundedTotalProvider.php]: Honestly, by this class name, I am unable to tell what it exactly does. I guess that the `repository` word is unnecessary.

File: src/Request/RefundUnitsRequest.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Request;
+
+use Sylius\RefundPlugin\Command\RefundUnits;
+use Symfony\Component\HttpFoundation\Request;
+use Webmozart\Assert\Assert;
+
+final class RefundUnitsRequest
```
Comment (bartoszpietrzak1994) [inline src/Request/RefundUnitsRequest.php]: `RefundUnitCommandProvider` ?

File: src/Resources/translations/flashes.en.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+sylius_refund:
+    at_least_one_unit_should_be_selected_to_refund: 'At least one unit should be selected to refund'
+    units_successfully_refunded: "Order units have been successfully refunded"
```
Comment (bartoszpietrzak1994) [inline src/Resources/translations/flashes.en.yml]: Single quotation

File: src/Resources/config/doctrine/Refund.orm.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+
+<doctrine-mapping
+        xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping
+                            http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd"
+>
+
+    <entity name="Sylius\RefundPlugin\Entity\Refund" table="sylius_refund_refund">
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Resources/config/doctrine/Refund.orm.xml]: Yeah, but we need to have the proper prefix + entity name :/

File: spec/CommandHandler/RefundUnitsHandlerSpec.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+        $eventBus->dispatch(Argument::that(function(UnitsRefunded $event): bool {
+            return
+                $event->orderNumber() === '000222' &&
+                $event->unitIds() === [1, 3] &&
+                $event->amount() === 1500
+            ;
+        }))->shouldBeCalled();
+
+        $this->__invoke(new RefundUnits('000222', [1, 3]));
```
Comment (GSadee) [inline spec/CommandHandler/RefundUnitsHandlerSpec.php]: Maybe `$this(new RefundUnits('000222', [1, 3]));`

File: src/Action/RefundUnitsAction.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        $this->session = $session;
+        $this->router = $router;
+    }
+
+    public function __invoke(Request $request): Response
+    {
+        try {
+            $this->commandBus->dispatch(RefundUnitsRequest::getCommand($request));
+
+            $this->session->getFlashBag()->add('success', 'sylius_refund.units_successfully_refunded');
```
Comment (GSadee) [inline src/Action/RefundUnitsAction.php]: I'm not sure if that is a right place to add notification. Maybe we should create a notification projector that will be listening to the `UnitsRefunded` event?

File: src/Factory/RefundFactoryInterface.php [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Factory;
+
+interface RefundFactoryInterface
+{
+    public function createWithData(string $orderNumber, int $unitId, int $amount);
```
Comment (GSadee) [inline src/Factory/RefundFactoryInterface.php]: Missing return type

File: src/Resources/config/doctrine/Refund.orm.xml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+
```
Comment (GSadee) [inline src/Resources/config/doctrine/Refund.orm.xml]: 😱

File: src/Resources/config/services.xml [hunk 10]
```diff
@@ -8,5 +8,50 @@
...
+            <argument type="service" id="sylius_refund.repository.refund" />
+        </service>
+
+        <service id="Sylius\RefundPlugin\Twig\OrderRefundsExtension">
+            <argument type="service" id="Sylius\RefundPlugin\Provider\OrderRefundedTotalProvider" />
+            <argument type="service" id="Sylius\RefundPlugin\Checker\UnitRefundingAvailabilityChecker" />
+            <tag name="twig.extension"/>
+        </service>
+
+        <!--Command Bus-->
```
Comment (GSadee) [inline src/Resources/config/services.xml]: So maybe move that to a separate file?

File: src/Resources/views/orderRefunds.html.twig [hunk 11]
```diff
@@ -43,39 +44,48 @@
...
+                                <th class="wide sylius-table-column-item">{{ 'sylius.ui.order_item_product'|trans }}</th>
+                                <th class="center aligned sylius-table-column-total">{{ 'sylius.ui.total'|trans }}</th>
+                                <th class="center aligned"></th>
                             </tr>
+                        </thead>
+                        <tbody>
+                        {% for item in order.items %}
+                            {% for unit in item.units %}
+                                {% set variant = item.variant %}
+                                {% set product = item.variant.product %}
```
Comment (GSadee) [inline src/Resources/views/orderRefunds.html.twig]: `{% set product = variant.product %}` 🎉

File: tests/Behat/Page/OrderRefundsPage.php [hunk 12]
```diff
@@ -15,6 +16,36 @@ public function getRouteName(): string
 
     public function countRefundableUnitsWithProduct(string $productName): int
     {
-        return count($this->getDocument()->findAll('css', sprintf('#refunds .unit:contains("%s")', $productName)));
+        return count($this->getUnitsWithProduct($productName));
+    }
+
+    public function getRefundedTotal(): string
+    {
+        return str_replace('Refunded total: ', '', $this->getDocument()->find('css', '#refundedTotal')->getText());
```
Comment (GSadee) [inline tests/Behat/Page/OrderRefundsPage.php]: `$this->getElement('refunded_total')->getText()`

File: src/Resources/views/orderRefunds.html.twig [hunk 13]
```diff
@@ -43,39 +44,48 @@
...
-                    </tbody>
-                </table>
+                        </tbody>
+                    </table>
+
+                    <button class="ui primary button" type="submit">{{ 'sylius.ui.refund'|trans }}</button>
+
+                    <br/><br/>
+
+                    <h3 id="refundedTotal">
```
Comment (GSadee) [inline src/Resources/views/orderRefunds.html.twig]: `id="refunded-total"`

**Comment by GSadee** [comment]
Thanks Mati! 🏅

**Pull request merged**